### PR TITLE
Allow package listing to be sorted by Downloads or Name

### PR DESCRIPTION
**Pull request by parkerl**
@ericmj how does this look?

  - Issue #107

### CODE REVIEW

**Comment by parkerl** [comment]
@ericmj I just took a peek at the issue and realized you wanted it to default to being sorted by downloads. My bad. How does it look other than that?

File: lib/hex_web/package.ex [hunk 1]
```diff
@@ -202,6 +203,20 @@ defmodule HexWeb.Package do
   defp like_escape(string, escape) do
     String.replace(string, escape, "\\\\\\1")
   end
+
+  defp sort(query, nil) do
+    sort(query, :downloads)
+  end
```
Comment (ericmj) [inline lib/hex_web/package.ex]: Looks like we don't need this clause?

File: lib/hex_web/package.ex [hunk 2]
```diff
@@ -135,9 +135,10 @@ defmodule HexWeb.Package do
     |> HexWeb.Repo.delete_all
   end
 
-  def all(page, count, search \\ nil) do
+  def all(page, count, search \\ nil, sort \\ :downloads) do
```
Comment (ericmj) [inline lib/hex_web/package.ex]: The main packages should sort by name, only searches should sort by downloads.

File: lib/hex_web/package.ex [hunk 3]
```diff
@@ -202,6 +203,20 @@ defmodule HexWeb.Package do
...
+  end
+
+  defp sort(query, :name) do
+    from d in query, order_by: :name
+  end
+
+  defp sort(query, :downloads) do
+    from p in query,
+    left_join: d in HexWeb.Stats.PackageDownload, on: p.id == d.package_id and d.view == "all",
+    order_by: [asc: is_nil(d.downloads), desc: d.downloads]
```
Comment (ericmj) [inline lib/hex_web/package.ex]: Beautiful :)

File: lib/hex_web/util.ex [hunk 4]
```diff
@@ -20,6 +20,12 @@ defmodule HexWeb.Util do
     :calendar.gregorian_days_to_date(today_days - 1)
   end
 
+  def last_year do
+    {today, _time} = :calendar.universal_time()
+    today_days = :calendar.date_to_gregorian_days(today)
+    :calendar.gregorian_days_to_date(today_days - 35)
+  end
```
Comment (ericmj) [inline lib/hex_web/util.ex]: This is only 35 days back, so it shouldn't be called last_year. It's only used in the sample_data script so move it there.

File: lib/hex_web/package.ex (hunk 1 above)
Comment (ericmj) [inline lib/hex_web/package.ex]: I see from the code below that we actually need this clause. Change the code below so that `nil => :name`.

**Comment by ericmj** [comment]
Looks good, left a few comments and I'm going to try it out locally.

**Comment by parkerl** [comment]
Changes made

File: lib/hex_web/web/router.ex [hunk 5]
```diff
@@ -186,6 +187,18 @@ defmodule HexWeb.Web.Router do
     |> String.strip
   end
 
+  defp safe_sort(nil), do: nil
```
Comment (ericmj) [inline lib/hex_web/web/router.ex]: This clause is not needed if you have the catch-all below.

File: lib/hex_web/package.ex [hunk 6]
```diff
@@ -135,9 +135,10 @@ defmodule HexWeb.Package do
     |> HexWeb.Repo.delete_all
   end
 
-  def all(page, count, search \\ nil) do
+  def all(page, count, search \\ nil, sort \\ :name) do
     from(p in HexWeb.Package,
-         order_by: p.name)
+         preload: :downloads)
+    |> sort(sort || :name)
```
Comment (ericmj) [inline lib/hex_web/package.ex]: Don't default here.

File: lib/hex_web/web/router.ex [hunk 7]
```diff
@@ -100,14 +100,15 @@ defmodule HexWeb.Web.Router do
     conn              = fetch_params(conn)
     packages_per_page = @packages
     search            = conn.params["search"] |> safe_query
+    sort              = conn.params["sort"] |> safe_sort
```
Comment (ericmj) [inline lib/hex_web/web/router.ex]: This is where you should default. Sort on :name by default and on :downloads if we are doing a search.

**Comment by jcspencer** [comment]
This looks good to me :+1:

The only question I have is whether packages should be sorted by a drop-down _sort by_ box instead of clicking the table headers? I personallythink it would look cleaner.

File: lib/hex_web/web/router.ex (hunk 7 above)
Comment (parkerl) [inline lib/hex_web/web/router.ex]: Ah of course. Sorry I was in too much of a hurry this afternoon. Thanks for your patience.

**Comment by ericmj** [comment]
@parkerl Great work. Please squash into a single commit and I will merge after.

@jcspencer Not sure where it would fit, feel free to play around it with after we merge.

**Comment by parkerl** [comment]
Done.

**Comment by jcspencer** [comment]
@ericmj I've put this together locally. What do you think?
If you're interested, I can open a PR.

screen shot 2015-05-16 at 3 16 35 pm
screen shot 2015-05-16 at 3 16 45 pm

**Comment by ericmj** [comment]
Beautiful work as usual James, please open a PR after the current one is
merged, should be by the end of the day.

**Pull request merged**